WIP: Refactor everything

anthonydavidson opened the pull request:
This is a work in progress.  Radio receive is not functional.

Changes:
* Added a start up sequence that's a rainbow strobe.
* There's no longer a state where you aren't running one of the expected animations (what was in `runAdjustments()`)
* Created a `MasterController` that `loop()` delegates to. It controls what mode the scarf is in - Off Mode or Animation Mode.
* Off Mode and Animation Mode each have their own controller. The MasterController delegates to one of them depending on which mode is active.
* Switching between modes is now done with a long click.
* Created `MilliTimer` class to capture concept of checking on time passed.
* All animations and Off Mode now only write to the LEDs when the lights need to be updated. I hope this increases battery life a small bit.
* Replaced click button part of `MyKnob` with `ButtonControl`. It's a singleton like MyKnob was.
* No longer have to release button for long click. An event is created once you've been holding it down long enough.
* An event is created for double click but it currently isn't used.
* Replaced rotary part of `MyKnob` with `KnobSetting`, though MyKnob was a singleton that had a bunch of references to it while KnobSetting captures (and contains) the information needed for a particular setting on an animation while still using a global `Encoder`.
* The state of a setting for a particular animation (that's altered by rotating the knob) is now saved and will be what's used by that animation when you return to it.
* Radio now sends only once a second and not when state of animation changes. Radio sending was so chatty because  the rotary knob doesn't have the equivalent of a debouncer. In other terms, the potentiometer that the rotary know is connected to increments more than once for each click of the rotary. This means that turning the knob to the next setting causes multiple state changes which are captured by the loop and sent out by the radio.

tony-zebra committed:
replaced off mode with strobe mode

anthonydavidson committed:
bunch of refactoring

anthonydavidson committed:
bunch of refactoring 2

anthonydavidson committed:
clean up

anthonydavidson committed:
clean up

counterbeing commented:
Hey there! Thanks for the PR.

I've just started to go over it. I'm totally down to merge it once the radios work :) and we have equivalent functionality.

Just for my understanding, as I'd brand myself as a C++ novice/hobbyist, I'd love to hear about why you changed what you did. 

Things definitely look a bit more succinct than before! I have to say that in writing a lot of the code I felt like there was a lot of duplication, but came to believe that was just a necessary evil of writing C++.

I may sprinkle a few questions in, but would love to hear about your overarching design ideas, or any obvious problems you saw in the code before.

Thanks!

counterbeing commented on src/main.cpp:
```diff
@@ -16,127 +18,157 @@
+// const int NUM_ANIMATONS = 2;
+Animation* animations[NUM_ANIMATONS] = {
+    // &crossfade, &color_chooser};
+    &crossfade, &color_chooser,    &race,  &stars, &rainbow, &fuck_my_eyes,
```
I love how much more succinct this is 👍

anthonydavidson replied:
Thanks!

counterbeing commented on src/animations/Rainbow.h:
```diff
@@ -2,24 +2,25 @@
+ public:
+  Rainbow(CRGB leds_[]) : leds(leds_) {}
+
+  void wakeUp() {
```
Why do we call these functions `wakeUp`?

anthonydavidson replied:
I changed this to `activate()`. Regardless, I was looking for a good name that wasn't `setup()`. I didn't want to reuse it because (AFAIK) the `setup()` method in `main.cpp` is only called once, while this is method is going to be called every time we go back to a particular animation. It's a minor distinction, but it made sense to me. I'm also open to calling it something else.

counterbeing replied:
Your thinking makes sense. I guess I was just following the existing pattern, but you're right, it's not an exact match. I do prefer `activate`.  👍

counterbeing commented on src/main.cpp:
```diff
@@ -16,127 +18,157 @@
+ public:
+  void setAsActive() { activeFlag = true; }
+
+  virtual void run() {
```
What do these virtual functions get us in this app?

anthonydavidson replied:
Virtual methods go on base classes and are likely to be overridden in derived classes. This helps make sure that the derived class's method gets called when what you have it a pointer to an instance of a derived class but the pointer itself has the type of the base class.  A better explanation is [here](https://www.geeksforgeeks.org/virtual-function-cpp/).

Basically, I use virtual if I know I'm going to override that method in a derived class. By always using virtual it acts like Java does by default, which I know a lot better than C++.

counterbeing replied:
Good explanation! Thanks!

anthonydavidson committed:
move leds member into base Animation class

anthonydavidson committed:
streamlined button events; off mode now uses long click

anthonydavidson committed:
simplified knob configured settings; added startup sequence

anthonydavidson committed:
animations extend base controller

anthonydavidson committed:
collapse KnobControl into KnobSetting

anthonydavidson committed:
include cleanup

anthonydavidson committed:
include cleanup

anthonydavidson commented:
Hey @counterbeing ! Thanks for your comments. My intent was to get this a little farther along (to where it is now, more or less) and then add in some explanations, which hopefully I'll get to tomorrow. I'm also happy to get on a call and walk you (& Mac) through any or all of it. Radio work still to come.

counterbeing commented:
Thanks @anthonydavidson! This all sounds perfect. I'd love to hop on the phone with you and Mac at some point just to do a code review so that I can learn a bit more.

I think in the coming weeks I'll be attempting to run a version of the software on new hardware (esp32) to see if I can adapt it. It's probably just easier hardware to run on. I'll try and run your new code if it's up!

Are you running your changes on a nano currently?

Cheers!

anthonydavidson committed:
startup sequence fix

anthonydavidson committed:
fixed radio send; ALSO BROKEN LEDS ON ANIMATION LOOP (hangs when animationIndex goes back to zero)

anthonydavidson committed:
fixed animationIndex reset bug

anthonydavidson commented:
Cory - Yes, I'm running on one of @100ideas 's scarves. 

I've got radio send working now, still need to update state on radio receive, but that should be it.

counterbeing committed:
🤖 set up clang-formatter

If this is trouble feel free to revert it, i just think it might save us
some pain later.

anthonydavidson committed:
radio receiving now works; made constants const
